lib/rust/data-structures/src/unrolled_linked_list.rs

```diff
@@ -0,0 +1,1110 @@
+}
+
+impl<T, const N: usize, B> Default for Node<T, N, B>
+where B: AllocationBehavior<T>
+{
+    #[inline(always)]
+    fn default() -> Self {
+        AllocationBehavior::new_node()
+    }
+}
+
+impl<T, const N: usize, B> Node<T, N, B> {
+    /// Count of all nodes starting at this node, including this node.
+    #[inline(always)]
+    fn node_count(&self) -> usize {
+        self.next.opt_item().map(|t| t.node_count()).unwrap_or(0) + 1
+    }
+
+    #[inline(always)]
+    fn shrink_to_fit(&mut self, count: usize) -> usize {
+        if count < N {
+            self.next.set_default();
+            N
+        } else {
+            self.next.opt_item_mut().unwrap().shrink_to_fit(count - N)
+        }
+    }
+
+    /// Get a mutable reference to items and the next node if it exists.
+    #[inline(always)]
+    fn items_and_next_mut(&mut self) -> (&mut Vec<NodeItem<T>>, Option<&mut Box<Self>>) {
+        let next = self.next.opt_item_mut();
+        // # Safety
+        // Self is mutably borrowed, so this call is safe.
+        #[allow(unsafe_code)]
+        let items = unsafe { self.items.unchecked_borrow_mut() };
+        (items, next)
+    }
+
+    #[inline(always)]
+    fn items(&self) -> &Vec<NodeItem<T>> {
+        // # Safety
+        // The only function that does not require mutable self access which mutably borrows items
+        // is [`Self::push_internal`], but it borrows newly created item only.
+        #[allow(unsafe_code)]
+        unsafe {
+            self.items.unchecked_borrow()
+        }
+    }
+
+    #[inline(always)]
+    fn items_mut(&mut self) -> &mut Vec<NodeItem<T>> {
+        // # Safety
+        // The only function that does not require mutable self access which mutably borrows items
+        // is [`Self::push_internal`], but it borrows newly created item only.
+        #[allow(unsafe_code)]
+        unsafe {
+            self.items.unchecked_borrow_mut()
+        }
+    }
+}
+
+impl<T, const N: usize, B> Node<T, N, B>
+where B: AllocationBehavior<T>
+{
+    #[inline(always)]
+    fn push(&self, offset: usize, elem: T) -> bool {
+        AllocationBehavior::push_to_node(self, offset, elem)
+    }
+
+    #[inline(always)]
+    fn push_internal(&self, offset: usize, elem: T, resize_items: bool) -> bool {
+        if offset < N {
+            if resize_items {
+                // # Safety
+                // The vector was initialized with capacity of `N` elements, so it is safe to set
+                // its new size here, as the value will be written below.
+                #[allow(unsafe_code)]
+                unsafe {
+                    self.items.unchecked_borrow_mut().set_len(offset + 1);
+                }
+            }
+
+            // # Safety
+            // The item at `offset` either does not exist yet or was removed, so there are no
+            // mutable references to it.
+            #[allow(unsafe_code)]
+            unsafe {
+                *self.items()[offset].unchecked_borrow_mut() = elem
```

Comment:
In case of `prealloc::Disabled`, the memory is uninitialized, right? So this deref would be UB.

Reply:
Ah, snap, I'm forgetting that deref of non-initialized memory is UB even if we are writing to it :( Btw, what are the reasons of that?